cmip6_downscaling/methods/common/tasks.py

```diff
@@ -272,11 +267,106 @@ def regrid(source_path: UPath, target_grid_path: UPath) -> UPath:
 
     regridder = xe.Regridder(source_ds, target_grid_ds, "bilinear", extrap_method="nearest_s2d")
     regridded_ds = regridder(source_ds)
+    regridded_ds.attrs.update(**get_cf_global_attrs(version=version))
     regridded_ds.to_zarr(target, mode='w')
 
     return target
 
 
+def _load_coords(ds: xr.Dataset) -> xr.Dataset:
+    '''Helper function to explicitly load all dataset coordinates'''
+    for var, da in ds.coords.items():
+        ds[var] = da.load()
+    return ds
+
+
+def _pyramid_postprocess(dt: dt.DataTree, levels: int, other_chunks: dict = None) -> dt.DataTree:
+    '''Postprocess data pyramid
+
+    Adds multiscales metadata and sets Zarr encoding
+
+    Parameters
+    ----------
+    dt : dt.DataTree
+        Input data pyramid
+    levels : int
+        Number of levels in pyramid
+    other_chunks : dict
+        Chunks for non-spatial dims
+
+    Returns
+    -------
+    dt.DataTree
+        Updated data pyramid with metadata / encoding set
+    '''
+    chunks = {"x": PIXELS_PER_TILE, "y": PIXELS_PER_TILE}
+    if other_chunks is not None:
+        chunks.update(other_chunks)
+
+    for level in range(levels):
+        slevel = str(level)
+        dt.ds.attrs['multiscales'][0]['datasets'][level]['pixels_per_tile'] = PIXELS_PER_TILE
+
+        # set dataset chunks
+        dt[slevel].ds = dt[slevel].ds.chunk(chunks)
+        if 'date_str' in dt[slevel].ds:
+            dt[slevel].ds['date_str'] = dt[slevel].ds['date_str'].chunk(-1)
+
+        # set dataset encoding
+        dt[slevel].ds = set_zarr_encoding(
+            dt[slevel].ds, codec_config={"id": "zlib", "level": 1}, float_dtype="float32"
+        )
+        for var in ['time', 'time_bnds']:
+            if var in dt[slevel].ds:
+                dt[slevel].ds[var].encoding['dtype'] = 'int32'
+
+    # set global metadata
+    dt.ds.attrs.update(**get_cf_global_attrs(version=version))
+    return dt
+
+
+@task(log_stdout=True, tags=['dask-resource:TASKSLOTS=1'])
+def pyramid(ds_path: UPath, levels: int = 2, other_chunks: dict = None) -> UPath:
+    '''Task to create a data pyramid from an xarray Dataset
+
+    Parameters
+    ----------
+    ds_path : UPath
+        Path to input dataset
+    levels : int, optional
+        Number of levels in pyramid, by default 2
+    uri : str, optional
+        Path to write output data pyamid to, by default None
+    other_chunks : dict
+        Chunks for non-spatial dims
+
+    Returns
+    -------
+    target : UPath
+    '''
+
+    target = intermediate_dir / "pyramid" / ds_path.path.replace('/', '_')
```

Comment:
do we want target to be `intermediate_dir` or `results_dir` (or something else? `pyramid_dir`?)